### PR TITLE
Implementation of a new chat

### PR DESCRIPTION
**Pull request by arek1-618**
Implementation of a new chat.

chrome_2018-10-11_07-08-24
chrome_2018-10-11_07-08-09

### CODE REVIEW

File: modules/Chat/actions/Entries.php [hunk 1]
```diff
@@ -20,9 +23,32 @@ class Chat_Entries_Action extends \App\Controller\Action
 	 */
 	public function checkPermission(\App\Request $request)
 	{
-		$currentUserPriviligesModel = Users_Privileges_Model::getCurrentUserPrivilegesModel();
-		if (!$currentUserPriviligesModel->hasModulePermission($request->getModule())) {
-			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED', 406);
+		if (
+			!Users_Privileges_Model::getCurrentUserPrivilegesModel()->hasModulePermission($request->getModule()) ||
+			!\App\Module::isModuleActive('Chat')
```
Comment (rskrzypczak) [inline modules/Chat/actions/Entries.php]: hasModulePermission function contains isModuleActive

File: modules/Chat/actions/Entries.php [hunk 2]
```diff
@@ -20,9 +23,32 @@ class Chat_Entries_Action extends \App\Controller\Action
...
+		) {
+			throw new \App\Exceptions\NoPermittedToRecord('ERR_NO_PERMISSIONS_FOR_THE_RECORD', 406);
```
Comment (rskrzypczak) [inline modules/Chat/actions/Entries.php]: Wrong throw

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 3]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
-	registerChat: function () {
-		const self = this;
-		var modal = $('.chatModal');
-		if (modal.length === 0) {
+	/**
+	 * Add new item to room list.
+	 * @param {int} roomId
+	 * @param {string} name
+	 */
+	addRoomItem(roomId, name) {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: relocate to chat.js

File: modules/Chat/actions/Entries.php [hunk 4]
```diff
@@ -3,9 +3,12 @@
...
+ * @package   Action
+ *
  * @copyright YetiForce Sp. z o.o
- * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
- * @author Mariusz Krzaczkowski <m.krzaczkowski@yetiforce.com>
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Mariusz Krzaczkowski <m.krzaczkowski@yetiforce.com>
+ * @author    Arkadiusz Adach <a.adach@yetiforce.com>
  */
 class Chat_Entries_Action extends \App\Controller\Action
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/Entries.php]: split into smaller files

File: modules/Chat/actions/Entries.php [hunk 5]
```diff
@@ -20,8 +23,26 @@ class Chat_Entries_Action extends \App\Controller\Action
...
+			($mode === 'addRoom' && !$request->has('record')) ||
+			($mode === 'switchRoom' && !$request->has('chat_room_id'))
+		) {
+			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED', 406);
+		}
+		if (
+			$request->has('record') &&
+			!Vtiger_Record_Model::getInstanceById($request->getInteger('record'))->isViewable()
+		) {
+			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED', 406);
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/Entries.php]: LBL_PERMISSION_DENIED  >> ERR_.....

File: modules/Chat/actions/Entries.php [hunk 6]
```diff
@@ -20,8 +23,26 @@ class Chat_Entries_Action extends \App\Controller\Action
...
+		}
+		if (
+			$request->has('chat_room_id') &&
+			$request->getInteger('chat_room_id') !== 0 &&
+			!Vtiger_Record_Model::getInstanceById($request->getInteger('chat_room_id'))->isViewable()
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/Entries.php]: Vtiger_Record_Model::getInstanceById  save to a variable

File: modules/Chat/actions/Entries.php [hunk 7]
```diff
@@ -32,18 +53,63 @@ public function checkPermission(\App\Request $request)
...
 	 * Add entries function.
 	 *
 	 * @param \App\Request $request
 	 */
-	public function add(\App\Request $request)
+	public function addMessage(\App\Request $request)
+	{
+		$room = \App\Chat::getInstanceById($request->getInteger('chat_room_id'));
+		$isAssigned = $room->isAssigned();
+		$room->addMessage($request->get('message'));
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/Entries.php]: $request->get() is forbidden

File: modules/Chat/views/Entries.php [hunk 8]
```diff
@@ -22,12 +25,47 @@ public function __construct()
...
+	 * Get HTML.
+	 *
+	 * @param \App\Request   $request
+	 * @param \App\Chat|null $room
+	 *
+	 * @throws \App\Exceptions\IllegalValue
+	 *
+	 * @return \html|string
+	 */
+	public function getHTML(\App\Request $request, \App\Chat $room = null)
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/views/Entries.php]: getHTML  >> getEntriesView

File: modules/Vtiger/views/Basic.php [hunk 9]
```diff
@@ -49,7 +49,7 @@ public function preProcess(\App\Request $request, $display = true)
 			$viewer->assign('SEARCHED_MODULE', $selectedModule);
 		}
 		if (\App\Module::isModuleActive('Chat')) {
-			$viewer->assign('CHAT_ENTRIES', (new Chat_Module_Model())->getEntries());
+			$viewer->assign('CHAT', \App\Chat::getInstanceById(\App\Chat::getCurrentRoomId()));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/Basic.php]: unnecessary

File: modules/Vtiger/views/Detail.php [hunk 10]
```diff
@@ -1028,6 +1049,31 @@ public function showSocialMedia($request)
...
+	 * @param \App\Request $request
+	 *
+	 * @throws \App\Exceptions\IllegalValue
+	 * @throws \App\Exceptions\NoPermittedToRecord
+	 *
+	 * @return \html
+	 */
+	public function showChat(\App\Request $request)
+	{
+		$recordModel = Vtiger_Record_Model::getInstanceById($request->getInteger('record'));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/Detail.php]: $recordModel = $this->record->getRecord();

File: public_html/layouts/basic/modules/Vtiger/resources/Detail.js [hunk 11]
```diff
@@ -1651,6 +1651,7 @@ jQuery.Class("Vtiger_Detail_Js", {
 					delete urlAttributes.callback;
 				}
 				thisInstance.loadContents(url, urlAttributes).done(function (data) {
+					Vtiger_Header_Js.getInstance().registerChatEvents($('.js-chat-detail'));
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Detail.js]: separate file

File: public_html/layouts/basic/modules/Vtiger/resources/Detail.js [hunk 12]
```diff
@@ -2602,6 +2603,7 @@ jQuery.Class("Vtiger_Detail_Js", {
 	},
 	registerEvents: function () {
 		//this.triggerDisplayTypeEvent();
+		Vtiger_Header_Js.getInstance().registerChatEvents($('.js-chat-detail'));
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Detail.js]: separate file

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 13]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+		let template = $('.js-chat-modal .js-room-template');
+		if (template.length) {
+			let item = template.clone(false, false);
+			item.removeClass('hide');
+			item.removeClass('js-room-template');
+			item.removeClass('fontBold');
+			item.find('.row').html(name);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: js-...

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 14]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+	/**
+	 * Add new item to room list.
+	 * @param {int} roomId
+	 * @param {string} name
+	 */
+	addRoomItem(roomId, name) {
+		let template = $('.js-chat-modal .js-room-template');
+		if (template.length) {
+			let item = template.clone(false, false);
+			item.removeClass('hide');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 15]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+			item.removeClass('js-room-template');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 16]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+			item.removeClass('fontBold');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 17]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+			item.find('.row').html(name);
+			item.data('roomId', roomId);
+			$('.js-chat-modal .js-chat-rooms-list').append(item);
+			this.registerSwitchRoom($('.js-chat-modal'));
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: $('.js-chat-modal') once declared

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 18]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+		if (typeof chatRoomId !== 'undefined') {
+			const self = this;
+			AppConnector.request({
+				module: 'Chat',
+				action: 'Entries',
+				mode: 'addRoom',
+				record: chatRoomId,
+			}).done((data) => {
+				if (data && data.success) {
+					if (container.find('.js-container-button').hasClass('hide')) {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: container.find('.js-container-button')  once declared

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 19]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+						container.find('.js-container-button').removeClass('hide');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: container.find('.js-container-button') once declared

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 20]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+						container.find('.js-container-items').addClass('hide');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: container.find('.js-container-button') once declared

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 21]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+					} else {
+						container.find('.js-container-button').addClass('hide');
+						container.find('.js-container-items').removeClass('hide');
+					}
+					this.addRoomItem(data.result.chat_room_id, data.result.name);
+				}
+			}).fail((error, err) => {
+				app.errorLog(error, err);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: remove

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 22]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+				}
+			}).done((dataResult) => {
+				self.updateChat(container, chatRoomId, dataResult.result.html);
+				inputMessage.val("");
+				icon.css('color', '#000');
+				if (dataResult.result['user_added_to_room']) {
+					self.addRoomItem(dataResult.result['room']['room_id'], dataResult.result['room']['name']);
+				}
+			}).fail((error, err) => {
+				app.errorLog(error, err);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: remove

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 23]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
-			clearTimeout(self.chatTimer);
+		}
+	},
+	/**
+	 * Register switch room.
+	 * @param {jQuery} container
+	 */
+	registerSwitchRoom(container) {
+		container.find('.js-change-room').off('click').on('click', (e) => {
+			let chatModal = $(e.currentTarget).closest('.js-chat-modal');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: once declared

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 24]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+			let roomId = $(e.currentTarget).data('roomId');
+			const progressIndicatorElement = $.progressIndicator({
+				'position': 'html',
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: position: 'html',

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 25]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+						container.find('.js-chat-items')
+							.removeClass('js-chat-room-' + prevChatRoomId)
+							.addClass('js-chat-room-' + roomId);
+					} else {
+						$(element).removeClass('fontBold');
+					}
+				});
+				progressIndicatorElement.progressIndicator({'mode': 'hide'});
+			}).fail((error, err) => {
+				app.errorLog(error, err);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: remove

File: public_html/layouts/basic/modules/Vtiger/resources/Header.js [hunk 26]
```diff
@@ -683,68 +684,251 @@ $.Class("Vtiger_Header_Js", {
...
+	registerChatCheck(timer, container) {
 		const self = this;
-		var icon = $('.chatModal .modal-title .fa-comments');
-		this.chatTimer = setTimeout(function () {
-			icon.css('color', '#00e413');
-			self.getChatItems();
-			self.registerChatLoadItems(timer);
-			icon.css('color', '#000');
+		self.chatCheckTimer = setTimeout(() => {
+			//self.getChatItems(container);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Header.js]: ?

**Pull request merged**